### PR TITLE
Vectorized: Use inline namespace instead of anonymous

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #67656
* __->__ #67655
* #68555

Some of the CPU operators already use the `namespace CPU_CAPABILITY` trick to avoid anonymous namespacing, like [`PowKernel.cpp`](https://github.com/pytorch/pytorch/blob/cd51d2a3ecc8ac579bee910f6bafe41a4c41ca80/aten/src/ATen/native/cpu/PowKernel.cpp#L14). This extends that pattern to the `Vectorized` class, which avoids `Wsubobject-linage` warnings like I was getting in #67621.

For many functions, it was necessary to add `inline` because the functions are defined in a header. There were no link errors previously because the anonymous namespace ensured they were not exposed to linkage. Similarly, free functions defined in an anonymous namespace might need the `C10_UNUSED` attribute to silence warnings about the function not being called in the only translation unit that it's defined in. By removing the anonymous namespace, these decorators are no longer necessary.

Differential Revision: [D32566109](https://our.internmc.facebook.com/intern/diff/D32566109)

### CODE REVIEW

**Comment by peterbell10** [comment]
@malfet PTAL

File: aten/src/ATen/cpu/vec/vec256/vec256.h [hunk 1]
```diff
@@ -28,29 +28,28 @@
...
- }
+// This header, and all of its subheaders, will be compiled with
+// different architecture flags for each supported set of vector
+// intrinsics. So we need to make sure they aren't inadvertently
+// linked together. We do this by declaring objects in an `inline
+// namespace` which changes the name mangling, but can still be
+// accessed as `at::vec`.
+inline namespace CPU_CAPABILITY {
+
+inline std::ostream& operator<<(std::ostream& stream, const c10::qint32& val) {
```
Comment (malfet) [inline aten/src/ATen/cpu/vec/vec256/vec256.h:41]: Can you please explain, why are you removing `C10_UNUSED` (I know that `inline` is sort of an equivalent, but `C10_UNUSED` has been validated across the plethora of compiler, so let's keep using it here)

File: aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h [hunk 2]
```diff
@@ -581,32 +581,32 @@ Vectorized<BFloat16> inline operator^(const Vectorized<BFloat16>& a, const Vecto
   return _mm256_xor_si256(a, b);
 }
 
-Vectorized<BFloat16> Vectorized<BFloat16>::eq(const Vectorized<BFloat16>& other) const {
+inline Vectorized<BFloat16> Vectorized<BFloat16>::eq(const Vectorized<BFloat16>& other) const {
```
Comment (malfet) [inline aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:584]: Please add the comment to PR description explaining why adding inline is necessary

File: aten/src/ATen/cpu/vec/vec_base.h [hunk 3]
```diff
@@ -843,25 +843,24 @@ inline mask_gather(const Vectorized<T>& src, T const* base_addr,
...
+    return Vectorized<dst_t>::loadu(static_cast<const void*>(src_arr));
+  }
+};
+
+template<typename scalar_t>
+struct CastImpl<scalar_t, scalar_t> {
+  static inline Vectorized<scalar_t> apply(const Vectorized<scalar_t>& src) {
+    return src;
+  }
+};
```
Comment (malfet) [inline aten/src/ATen/cpu/vec/vec_base.h:863]: Can you please move this change to a separate PR (as eliminating nested anonymous namespace is unrelated to this commit title)?

**Review by malfet — APPROVED** [review]
LGTM, but please address comments

File: aten/src/ATen/cpu/vec/vec256/vec256.h (hunk 1 above)
Comment (peterbell10) [reply aten/src/ATen/cpu/vec/vec256/vec256.h:41]: When the function was static, each compilation unit has it's own copy and the compiler warns that it is unused; so this macro was needed to silence the warning. But it's not needed any more  because `inline` functions are expected to not be used in every single compilation unit they're defined.

**Comment by peterbell10** [comment]
@malfet I've moved the nested namespace removal into a new PR and added a note in this PR's description about `inline` and `C10_UNUSED` changes.

**Comment by malfet** [comment]
@malfet has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32566109).

**Comment by malfet** [comment]
Internal build complains about:
```
aten/src/ATen/native/cpu/PowKernel.cpp:14:11: error: inline namespace reopened as a non-inline namespace [-Werror]
namespace CPU_CAPABILITY {
```
Will propose the change to this PR as well as do corresponding internal change
